lib/ansible/cli/doc.py

```diff
@@ -172,22 +371,93 @@ def display_plugin_list(self, results):
         # display results
         DocCLI.pager("\n".join(text))
 
+    def _display_available_roles(self, list_json):
+        """Display all roles we can find with a valid argument specification.
+
+        Output is: fqcn role name, entry point, short description
+        """
+        roles = list(list_json.keys())
+        entry_point_names = set()
+        for role in roles:
+            for entry_point in list_json[role]['entry_points'].keys():
+                entry_point_names.add(entry_point)
+
+        max_role_len = 0
+        max_ep_len = 0
+
+        if roles:
+            max_role_len = max(len(x) for x in roles)
+        if entry_point_names:
+            max_ep_len = max(len(x) for x in entry_point_names)
+
+        linelimit = display.columns - max_role_len - max_ep_len - 5
+        text = []
+
+        for role in sorted(roles):
+            for entry_point, desc in iteritems(list_json[role]['entry_points']):
+                if len(desc) > linelimit:
+                    desc = desc[:linelimit] + '...'
+                text.append("%-*s %-*s %s" % (max_role_len, role,
+                                              max_ep_len, entry_point,
+                                              desc))
+
+        # display results
+        DocCLI.pager("\n".join(text))
+
+    def _display_role_doc(self, role_json):
+        roles = list(role_json.keys())
+        text = []
+        for role in roles:
+            text += self.get_role_man_text(role, role_json[role])
+
+        # display results
+        DocCLI.pager("\n".join(text))
+
     def run(self):
 
         super(DocCLI, self).run()
 
+        basedir = context.CLIARGS['basedir']
         plugin_type = context.CLIARGS['type']
         do_json = context.CLIARGS['json_format']
+        roles_path = context.CLIARGS['roles_path']
+
+        if basedir:
+            AnsibleCollectionConfig.playbook_paths = basedir
+
+            # Add any 'roles' subdir in playbook dir to the roles search path
+            subdir = os.path.join(basedir, "roles")
+            if os.path.isdir(subdir):
+                roles_path = (subdir,) + roles_path
+
+        if plugin_type == 'role':
+            if context.CLIARGS['list_dir']:
+                # If an argument was given with --list, it is a collection filter
+                coll_filter = None
+                if len(context.CLIARGS['args']) == 1:
```

Comment:
More than 1 should be handled too

Reply:
I just meant it should either error or iterate. Right now you can pass multiple arguments and they're ignored. I didn't check what ansible-doc does for the rest of it, but I think erroring would be more clear.

Reply:
We now error in latest commit.